File: kouston-web/src/app/models/telemetry.ts
export interface Vessel {
  name: string;
  altitude: number;
  velocity: number;
  latitude: number;
  longitude: number;
  apoapsis: number;
  periapsis: number;
  fuel: number;
  timestamp: number;

  // Orbital elements
  inclination: number;
  eccentricity: number;
  semiMajorAxis: number;
  lan: number; // Longitude of Ascending Node
  argumentOfPeriapsis: number;
  trueAnomaly: number;
  period: number;
  bodyRadius: number;
  bodyName: string;
  atmosphereHeight: number; // Height of atmosphere in meters (0 = no atmosphere)

  // EDL data
  pitch: number;           // Pitch angle relative to horizon (degrees)
  heading: number;         // Attitude compass heading (degrees, 0 = north)
  prograde: number;        // Velocity compass heading (degrees, 0 = north)
  verticalSpeed: number;   // Vertical velocity (m/s, negative = descending)
  horizontalSpeed: number; // Horizontal velocity (m/s)
  radarAltitude: number;   // Height above terrain (m)
  throttle: number;        // Current throttle (0-1)

  // System bodies (flattened arrays)
  bodyNames: string[];
  bodyParents: string[];  // Parent body name (empty string = orbits Sun)
  bodyTrueAnomalies: number[];
  bodyArgsOfPeriapsis: number[];
  bodyLANs: number[];
  bodyInclinations: number[];
  bodySemiMajorAxes: number[];
  bodyEccentricities: number[];
  bodyRadii: number[];

  // Resource breakdown
  resourcePartNames: string[];
  resourceTypes: string[];
  resourceAmounts: number[];
  resourceMaxAmounts: number[];
}

export interface TelemetryState {
  [clientId: string]: Vessel;
}

File: server/telemetry/telemetry.go
package telemetry

import "sync"

type Vessel struct {
	Name      string  `json:"name"`
	Altitude  float64 `json:"altitude"`
	Velocity  float64 `json:"velocity"`
	Latitude  float64 `json:"latitude"`
	Longitude float64 `json:"longitude"`
	Apoapsis  float64 `json:"apoapsis"`
	Periapsis float64 `json:"periapsis"`
	Fuel      float64 `json:"fuel"`
	Timestamp int64   `json:"timestamp"`

	// Orbital elements
	Inclination        float64 `json:"inclination"`
	Eccentricity       float64 `json:"eccentricity"`
	SemiMajorAxis      float64 `json:"semiMajorAxis"`
	LAN                float64 `json:"lan"`
	ArgumentOfPeriapsis float64 `json:"argumentOfPeriapsis"`
	TrueAnomaly        float64 `json:"trueAnomaly"`
	Period             float64 `json:"period"`
	BodyRadius         float64 `json:"bodyRadius"`
	BodyName           string  `json:"bodyName"`
	AtmosphereHeight   float64 `json:"atmosphereHeight"`

	// EDL data
	Pitch           float64 `json:"pitch"`
	VerticalSpeed   float64 `json:"verticalSpeed"`
	HorizontalSpeed float64 `json:"horizontalSpeed"`
	RadarAltitude   float64 `json:"radarAltitude"`
	Throttle        float64 `json:"throttle"`

	// System bodies (flattened arrays)
	BodyNames           []string  `json:"bodyNames"`
	BodyTrueAnomalies   []float64 `json:"bodyTrueAnomalies"`
	BodyArgsOfPeriapsis []float64 `json:"bodyArgsOfPeriapsis"`
	BodyLANs            []float64 `json:"bodyLANs"`
	BodyInclinations    []float64 `json:"bodyInclinations"`
	BodySemiMajorAxes   []float64 `json:"bodySemiMajorAxes"`
	BodyRadii           []float64 `json:"bodyRadii"`

	// Resource breakdown
	ResourcePartNames  []string  `json:"resourcePartNames"`
	ResourceTypes      []string  `json:"resourceTypes"`
	ResourceAmounts    []float64 `json:"resourceAmounts"`
	ResourceMaxAmounts []float64 `json:"resourceMaxAmounts"`
}

type State struct {
	mu      sync.RWMutex
	Vessels map[string]*Vessel
}

func NewState() *State {
	return &State{
		Vessels: make(map[string]*Vessel),
	}
}

func (s *State) UpdateVessel(clientID string, v *Vessel) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.Vessels[clientID] = v
}

func (s *State) RemoveVessel(clientID string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	delete(s.Vessels, clientID)
}

func (s *State) GetAllVessels() map[string]*Vessel {
	s.mu.RLock()
	defer s.mu.RUnlock()

	result := make(map[string]*Vessel)
	for k, v := range s.Vessels {
		result[k] = v
	}
	return result
}
